vispy/visuals/filters/base_filter.py

```diff
@@ -120,3 +125,98 @@ def _detach(self, visual):
 
         self._attached = False
         self._visual = None
+
+
+class PrimitivePickingFilter(Filter, metaclass=ABCMeta):
+    """Abstract base class for Visual-specific filters to implement a
+    primitive-picking mode.
+
+    Subclasses must (and usually only need to) implement
+    :py:meth:`_update_id_colors`.
+    """
+
+    def __init__(self):
+        vfunc = Function("""\
+            varying vec4 v_marker_picking_color;
+            void prepare_marker_picking() {
+                v_marker_picking_color = $ids;
+            }
+        """)
+        ffunc = Function("""\
+            varying vec4 v_marker_picking_color;
+            void marker_picking_filter() {
+                if ($enabled != 1) {
+                    return;
+                }
+                gl_FragColor = v_marker_picking_color;
+            }
+        """)
+
+        self._id_colors = VertexBuffer(np.zeros((0, 4), dtype=np.float32))
+        vfunc['ids'] = self._id_colors
+        self._n_primitives = 0
+        super().__init__(vcode=vfunc, fcode=ffunc)
+        self.enabled = False
+
+    @abstractmethod
+    def _update_id_colors(self):
+        """Calculate the colors encoding the picking IDs for the visual.
+
+        Generally, this method should be implemented to:
+            1. Calculate the number of primitives in the visual, stored in
+            `self._n_primitives`.
+            2. Pack the picking IDs into a 4-component float array (RGBA
+            VertexBuffer), stored in `self._id_colors`.
```

Comment:
Would be awesome if the packing was done here instead of in the subclasses, since it's error prone and anyways it's always the same... Youy probbaly nee dto juggle some extra submethods calling each other, but it would be useful. not blocking though!

Reply:
I agree but didn't immediately have a way to do it that felt clean/intuitive so I punted and tried to compensate in the docstring. I'll give it a bit more thought because it does seem possible.

An alternative would be to move the packing into the shader and instead just pass the IDs in the `VertexBuffer`. I avoided this for now just because I'm more comfortable with numpy than I am with GLSL but I don't think it would be too difficult.

Since these are also "just" implementation details I'm also happy to explore both options in a new PR if we want this to (potentially) be available a bit quicker.

Reply:
See [36db4fc](https://github.com/vispy/vispy/pull/2500/commits/36db4fc96dbffd8293f6583e43595292542e3079) for one idea

Reply:
I was thinking more something like:

```python
@staticmethod
def _pack_ids_into_rgba(ids):
	# pack ids
	...
	return packed

@abstractmethod
def _get_picking_ids(self):
	# this is the method that subclasses will change, which simply
	# calculates the ids and spits them out as normal numpy array
	# and it could for example return None if no change is needed
	# to improve performance
	return None

def _update_id_colors(self):
	# this should remain untouched
	ids = self._get_picking_ids()
	if ids is None:
		return

	id_colors = self._pack_ids_into_rgba(ids)
	self._id_colors.set_data(id_colors)

def _on_data_updated(self, event=None):
	if not self.attached:
		return
	self._update_id_colors()
```

A bit more complicated, but removes the need to think about packing from the sublcasses, and should make it easier to switch to a shader implementation later as well.

Reply:
> Since these are also "just" implementation details I'm also happy to explore both options in a new PR if we want this to (potentially) be available a bit quicker.

To be clear, perfectly happy with this as well!